### Business Analytics - Assignment 1 {-}


---

**Assignment Points**: 100  
**Due Date**: Friday Week 4 (18 August) @ 11.59pm  
**Submission**: Submit the assignment using the submission link on iLearn


- Put **all your work** into this file;
- Solutions must be provided in the cells below each question. Do not insert new cells; 
- Failure to follow these instructions will result in a loss of 20 points.


### About the Assignment

Credit score cards are used as a risk control method in the financial industry. Personal information submitted by credit card applicants are used to predict the probability of future defaults. The bank employs such data to decide whether to issue a credit card to the applicant or not.




| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| Income   | Annual income  |  |
| Gender   | Applicant's Gender   | Male = 0, Female = 1  |
| Car | Car Ownership | Yes = 1, No = 0 | 
| Children | Number of Children | |
| Real Estate | Real Estate Ownership | Yes = 1, No = 0 
| Days Since Birth | No. of Days | Count backwards from current day (0), -1 means yesterday
| Days Employed | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| Payment Default | Whether a client has overdue credit card payments | Yes = 1, No = 0



---


### Problem 1 - (50 points) {-}


**Question 1** 

- Import the `assignment_data.xlsx` file from `data` folder into a pandas DataFrame named `df`; 
- Delete duplicate rows from `df` according to `ID`;
- Delete the `ID` column.
- How many rows are left in `df`? (answer in the Markdown box below in a full sentence)

(10 points)

In [1]:
import pandas as pd

df = pd.read_excel("C:/Users/asus/Desktop/BUSA8001/Assignment 1/data/assignment_data.xlsx")
df = df.drop_duplicates(subset='ID')
df = df.drop(columns='ID')
df

Gender  Car  Real Estate  Children    Income  Days Since Birth  \
0          1    1            1         1  135000.0            -10434   
1          1    0            0         0  157500.0            -22494   
2          1    0            1         0  675000.0            -17964   
3          1    0            0         0  193500.0            -17880   
4          0    1            0         0  270000.0            -20776   
...      ...  ...          ...       ...       ...               ...   
9993       1    1            0         0   81000.0            -17493   
9994       1    0            1         0   45000.0            -22072   
9995       1    0            0         0  306000.0            -12677   
9996       1    0            0         0  121500.0            -23491   
9998       1    0            1         0  144000.0            -19002   

      Days Employed  Payment Default  
0              -803                0  
1            365243                1  
2             -6594                0  
3             -9858                1  
4             -2745                1  
...             ...              ...  
9993          -2347                0  
9994         365243                0  
9995          -5587                1  
9996         365243                0  
9998         365243                0  

[6805 rows x 8 columns]

### <u>Answer</u>: **6805** rows remaining


---

**Question 2**

- Reset the index in `df` using an appropriate function from `pandas` so that the new index corresponds to the number of rows (make sure to delete the old index). 
- How many positive values of `Days Employed` are there? (answer in Markdown)
- Replace the positive values of `Days Employed` with 0 (zero) in `df`

(10 points)

In [2]:
df.reset_index(drop=True, inplace=True)

positive_values_count = df[df['Days Employed'] > 0]['Days Employed'].count()
positive_values_count

1069

### <u>Answer</u>: **1069** rows **positive**.


In [3]:
df['Days Employed'] = df['Days Employed'].apply(lambda 
                                                x:0 if x>0 
                                                else x)
new_positive_values_count = df[df['Days Employed'] > 0]['Days Employed'].count()
new_positive_values_count

0

---
**Question 3**

Create two new variables in `df` named 

1. `Age`;
2. `Years in Employment`,

which measure age and employment length in **years** (positive decimal numbers) from `Days Since Birth` and `Days Employed` by applying appropriate transformations on these variables. 

Delete the original variables `Days Since Birth` and `Days Employed`.

(5 points)


In [4]:
df['Age'] = abs(df['Days Since Birth']/365)
df['Years in Employment'] = abs(df['Days Employed']/365)

df.drop(['Days Since Birth','Days Employed'], axis=1, inplace=True)
df

Gender  Car  Real Estate  Children    Income  Payment Default  \
0          1    1            1         1  135000.0                0   
1          1    0            0         0  157500.0                1   
2          1    0            1         0  675000.0                0   
3          1    0            0         0  193500.0                1   
4          0    1            0         0  270000.0                1   
...      ...  ...          ...       ...       ...              ...   
6800       1    1            0         0   81000.0                0   
6801       1    0            1         0   45000.0                0   
6802       1    0            0         0  306000.0                1   
6803       1    0            0         0  121500.0                0   
6804       1    0            1         0  144000.0                0   

            Age  Years in Employment  
0     28.586301             2.200000  
1     61.627397             0.000000  
2     49.216438            18.065753  
3     48.986301            27.008219  
4     56.920548             7.520548  
...         ...                  ...  
6800  47.926027             6.430137  
6801  60.471233             0.000000  
6802  34.731507            15.306849  
6803  64.358904             0.000000  
6804  52.060274             0.000000  

[6805 rows x 8 columns]

---
**Question 4**

- Create a **one**-dimensional NumPy array named `y` by exporting the first 5,000 observations of `Payment_Default`. (Hint: see `ravel()` function)
- Create a NumPy array named `X` by exporting the first 5,000 observations of the following columns `Gender`, `Car`, `Real Estate`, `Children`, `Income`, `Age`, `Years in Employment`.
 
(10 points)


In [5]:
import numpy as np
y = df['Payment Default'].values[:5000]

x_columns = ['Gender', 'Car', 'Real Estate', 'Children', 'Income', 'Age', 'Years in Employment']
x = df[x_columns].values[:5000]
print(x)

[[1.00000000e+00 1.00000000e+00 1.00000000e+00 ... 1.35000000e+05
  2.85863014e+01 2.20000000e+00]
 [1.00000000e+00 0.00000000e+00 0.00000000e+00 ... 1.57500000e+05
  6.16273973e+01 0.00000000e+00]
 [1.00000000e+00 0.00000000e+00 1.00000000e+00 ... 6.75000000e+05
  4.92164384e+01 1.80657534e+01]
 ...
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 ... 9.90000000e+04
  4.89753425e+01 1.43534247e+01]
 [1.00000000e+00 0.00000000e+00 1.00000000e+00 ... 1.57500000e+05
  5.01726027e+01 1.35506849e+01]
 [1.00000000e+00 1.00000000e+00 1.00000000e+00 ... 2.25000000e+05
  4.23095890e+01 7.94520548e-01]]


---

**Question 5** 

- Use an appropriate `scikit-learn` library we learned in class to create the following NumPy arrays: `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% training and 30% test datasets. 
- Set `random_state` to 0 and stratify subsamples so that train and test datasets have roughly equal proportions of the target's class labels. 

(5 points) 

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(x, y, test_size = 0.3, random_state = 0, stratify=y)

print("Class distribution:")
print(pd.Series(y).value_counts(normalize=True))
print(" ")
print("Train class distribution:")
print(pd.Series(Y_train).value_counts(normalize=True))
print(" ")
print("Test class distribution:")
print(pd.Series(Y_test).value_counts(normalize=True))

Class distribution:
0    0.6128
1    0.3872
dtype: float64
 
Train class distribution:
0    0.612857
1    0.387143
dtype: float64
 
Test class distribution:
0    0.612667
1    0.387333
dtype: float64


---

**Question 6**

- Create new variables by using an appropriate `scikit-learn` library we learned in class to standardize the features from the training and test datasets to mean zero and variance one. Name the new variables by appending '_scaled' to the original variable names.


(10 points)   

In [7]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled_x_train = scaler.fit_transform(X_train)
scaled_x_test = scaler.transform(X_test)

scaled_col = [col + '_scaled' for col in x_columns]
scaled_x_train = pd.DataFrame(scaled_x_train, columns=scaled_col)
scaled_y_train = pd.DataFrame(scaled_x_test, columns=scaled_col)

print("Scaled Training set:")
print(scaled_x_train)
print(" ")
print("Scaled Testing set:")
print(scaled_x_test)

Scaled Training set:
      Gender_scaled  Car_scaled  Real Estate_scaled  Children_scaled  \
0         -1.403074    1.264265            0.733799         0.627702   
1          0.712721    1.264265            0.733799         1.779146   
2          0.712721    1.264265            0.733799        -0.523743   
3          0.712721   -0.790973            0.733799        -0.523743   
4         -1.403074    1.264265            0.733799        -0.523743   
...             ...         ...                 ...              ...   
3495      -1.403074   -0.790973            0.733799        -0.523743   
3496       0.712721   -0.790973           -1.362770         0.627702   
3497      -1.403074    1.264265            0.733799        -0.523743   
3498       0.712721   -0.790973           -1.362770         0.627702   
3499       0.712721   -0.790973            0.733799         1.779146   

      Income_scaled  Age_scaled  Years in Employment_scaled  
0         -0.486428    0.651963                    1

---

## Problem 2 - (20 Points) {-}

**Question 7**

Fit the following two classifiers to the transformed training dataset using `scikit-learn` libraries.

- Perceptron - name your instance `pc` set `random_state=7`
- Logistic Regression - name your instance `lr` set `random_state=7`

When initializing instances of the above classifiers only set the parameters referenced above and nothing else.

(20 points)

In [8]:
from sklearn.linear_model import Perceptron, LogisticRegression

pc=Perceptron(random_state=7)
lr=LogisticRegression(random_state=7)

pc.fit(scaled_x_train, Y_train)
lr.fit(scaled_x_train, Y_train)

LogisticRegression(random_state=7)

---
## Problem 3 - (30 points) {-}


**Question 8**

- Using a method built into each of the two classifiers compute their prediction accuracies on the training data;
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_train`, e.g. you should have `lr_accuracy_train`; 
- Print the two accuracy **variables** along with their brief descriptions.

(5 points)

In [9]:
from sklearn.metrics import accuracy_score

pc_predict_train = pc.predict(scaled_x_train)
lr_predict_train = lr.predict(scaled_x_train)

pc_acc_train = accuracy_score(Y_train, pc_predict_train)
lr_acc_train = accuracy_score(Y_train, lr_predict_train)

print("Training data perceptron accuracy:", pc_acc_train)
print("Training data logistic regression accuracy", lr_acc_train)

Training data perceptron accuracy: 0.5202857142857142
Training data logistic regression accuracy 0.6088571428571429


---

**Question 9** 

- Using a method built into each of the above classifiers compute their prediction accuracy for the test dataset
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_test`, e.g. you should have `lr_accuracy_test`. 
- Print the two accuracy **variables** along with brief descriptions.

(5 points)

In [10]:
pc_predict_test = pc.predict(scaled_x_test)
lr_predict_test = lr.predict(scaled_x_test)

pc_acc_test = accuracy_score(Y_test, pc_predict_test)
lr_acc_test = accuracy_score(Y_test, lr_predict_test)

print("Test data perceptron accuracy", pc_acc_test)
print("Test data logistic regression accuracy", lr_acc_test)

Test data perceptron accuracy 0.5026666666666667
Test data logistic regression accuracy 0.6173333333333333


C:\Users\asus\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but Perceptron was fitted with feature names
  warnings.warn(
C:\Users\asus\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


---

**Question 10** 

Using nicely formated text in Markdown comment on the accuracies computed in Questions 8 & 9 making sure you address:
1. Training vs test set datasets for each classifier; (5 marks)
2. Perceptrion vs Logistic Regression. (5 marks) 
3. Are the results as expected, and why or why not? (Hint: You are not expected to comment on why a particular model is better. 10 marks) 

(20 marks)


<h3 style="text-align: center;">Summary</h3>

| *Data*              | Perceptron accuracy| Logistic Regression accuracy |
| :---------------- | :--------: | :------------------: |
| **Training data**     |   52.03 %     |        60.89 %        |
| **Test data**         |   50.27 %     |        61.73 %        |


#### 1. Training vs test set datasets for each classifier
*On the test dataset, the accuracy scores for **both classifiers are lower** than on the **training dataset**. This states that models **perform better with data already used while training**.*

#### 2. Perceptrion vs Logistic Regression.
***Logistic Regression outweighs the Perceptron** in terms of **test accuracy**. In comparision to Perceptron, the **Logistic Regression classifier performs better on both the training and test datasets**. This suggests that Logistic Regression is better effective at capturing the underlying patterns in data and generalising.*

#### 3. Are the results as expected, and why or why not?

*The outcomes are as expected and show frequent tendencies in machine learning. The models learned certain underlying patterns from the training data, which resulted in reasonable accuracy, on both training and test data.* 

*On the other hand, the small variation in test accuracies, shows that further optimisation may be required for even better generalisation.*


<span style="font-family: Roboto; font-size: 25px"> ***- Jatin Jassi*** 
</span> <span style="font-family: Roboto; font-size: 25px"> *(46824014)* </span>

---
---

### Marking Criteria

- To receive full marks your solutions must satisfy the following criteria:
    - Provide numerically correct answers;
    - Use the specific Python methods and libraries asked for in each question;
    - Written answers (where required) must be correct and formulated in easy to understand full sentences. 

---
---